### PR TITLE
[W5.6a][W12-4] Jin Minjia

### PR DESCRIPTION
**Pull request by bluein-green**

### CODE REVIEW

**Comment by Lunastryke** [comment]
Good implementation of contact class.

**Comment by CT15** [comment]
Good job. I don't spot any issue with this PR.
I will close this PR.

**Comment by CT15** [comment]
Perhaps you can specify the purpose of the PR.
Also, do you think that `value` can be extracted out from `Address`, `Email` and `Phone` to their parent class?

File: src/seedu/addressbook/data/person/Contact.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package seedu.addressbook.data.person;
+
+import seedu.addressbook.data.exception.IllegalValueException;
+
+/**
+ * Represents a contact in contact list
```
Comment (CT15) [inline src/seedu/addressbook/data/person/Contact.java:6]: Does not have to be too specific on where the class is used (i.e. `in the contact list`). Make it more general. `Contact` does not always have to be bounded by a list.

File: src/seedu/addressbook/data/person/Contact.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+ */
+public class Contact {
```
Comment (CT15) [inline src/seedu/addressbook/data/person/Contact.java:8]: You can actually push the constructor code to `Contact` class (parent class).

So something like

```java
public Contact(String value, boolean isPrivate, String validation, String constraints) throws IllegalValueException) { /* initialise attributes ... */ }
```

File: src/seedu/addressbook/data/person/Contact.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    protected boolean isPrivate;
+
+    public boolean isPrivate() {
+        return isPrivate;
+    }
+}
```
Comment (CT15) [inline src/seedu/addressbook/data/person/Contact.java:14]: You can consider `@Override` both `equals()` and `hashcode()` methods for `Contact` class as well. Perhaps you want to find out why we may want to override these two methods.

File: src/seedu/addressbook/data/person/Email.java [hunk 4]
```diff
@@ -6,15 +6,14 @@
  * Represents a Person's email in the address book.
  * Guarantees: immutable; is valid as declared in {@link #isValidEmail(String)}
  */
-public class Email {
+public class Email extends Contact {
```
Comment (CT15) [inline src/seedu/addressbook/data/person/Email.java:9]: If you push the constructor to the superclass, you need to call `super()` method in the `Email` constructor (basically all child classes' constructors). The `super()` method is used to call superclass constructor.

So something like

```java
public Email(String email, boolean isPrivate) throws IllegalValueException {
        super(email, isPrivate, EMAIL_VALIDATION_REGEX, MESSAGE_EMAIL_CONSTRAINTS);
}
```

File: src/seedu/addressbook/data/person/Email.java [hunk 5]
```diff
@@ -6,15 +6,14 @@
...
 
     public static final String EXAMPLE = "valid@e.mail";
     public static final String MESSAGE_EMAIL_CONSTRAINTS =
             "Person emails should be 2 alphanumeric/period strings separated by '@'";
     public static final String EMAIL_VALIDATION_REGEX = "[\\w\\.]+@[\\w\\.]+";
 
     public final String value;
```
Comment (CT15) [inline src/seedu/addressbook/data/person/Email.java:16]: `value` is an attribute that is present in all child classes. Can consider moving this to superclass.

**Review by CT15 — COMMENTED** [review]
Hi I have done the review for this PR. If anything is not clear, you can reply on specific comments. Otherwise, you can close this PR.

Good job! 👍

**Pull request closed without merging**